src/record/partially_serialized.rs

```diff
@@ -0,0 +1,111 @@
+use crate::blob::File;
+use crate::prelude::*;
+use bytes::{Bytes, BytesMut};
+
+pub(crate) struct PartiallySerializedWriteResult {
+    bytes_written: u64,
+    header_checksum: u32,
+}
+
+impl PartiallySerializedWriteResult {
+    pub(crate) fn bytes_written(&self) -> u64 {
+        self.bytes_written
+    }
+
+    pub(crate) fn header_checksum(&self) -> u32 {
+        self.header_checksum
+    }
+}
+
+pub(crate) struct PartiallySerializedRecord {
+    head_with_data: BytesMut,
+    header_len: usize,
+    data: Option<Bytes>,
+}
+
+impl PartiallySerializedRecord {
+    pub(crate) fn new(head_with_data: BytesMut, header_len: usize, data: Option<Bytes>) -> Self {
+        Self {
+            head_with_data,
+            header_len,
+            data,
+        }
+    }
+
+    pub(super) fn serialize_with_checksum(
+        self,
+        blob_offset: u64,
+    ) -> Result<(Bytes, Option<Bytes>, u32)> {
+        let Self {
+            head_with_data,
+            header_len,
+            data,
+        } = self;
+        let (head, checksum) =
+            Self::finalize_with_checksum(head_with_data, header_len, blob_offset)?;
+        Ok((head.freeze(), data, checksum))
+    }
+
+    pub(crate) async fn write_to_file(
+        self,
+        file: &File,
+        offset: u64,
+    ) -> Result<PartiallySerializedWriteResult> {
+        let (head, data, checksum) = self.serialize_with_checksum(offset)?;
+
+        let res = if let Some(data) = data {
+            Self::write_double_pass(head, data, file).await
+        } else {
+            Self::write_single_pass(head, file).await
+        };
+
+        res.map(|l| PartiallySerializedWriteResult {
+            bytes_written: l,
+            header_checksum: checksum,
+        })
+    }
+
+    fn finalize_with_checksum(
+        mut buf: BytesMut,
+        header_len: usize,
+        bytes_offset: u64,
+    ) -> Result<(BytesMut, u32)> {
+        use std::mem::size_of;
+
+        let offset_pos = RecordHeader::blob_offset_offset(header_len);
+        let checksum_pos = RecordHeader::checksum_offset(header_len);
+        let offset_slice = &mut buf[offset_pos..];
+        offset_slice[..size_of::<u64>()].copy_from_slice(&bytes_offset.to_le_bytes());
+
+        let checksum_slice = &mut buf[checksum_pos..];
+        checksum_slice[..size_of::<u32>()].copy_from_slice(&0u32.to_le_bytes());
+
+        let header_slice = &buf[..header_len];
+        let checksum: u32 = CRC32C.checksum(header_slice);
+        let checksum_slice = &mut buf[checksum_pos..];
+        checksum_slice[..size_of::<u32>()].copy_from_slice(&checksum.to_le_bytes());
```

Comment:
You can reuse `checksum_slice` created above

Reply:
No, since it's mutable borrow can't cross immutable borrow in header slice